File: apps/web/src/features/notifications/api/mappers.ts
/**
 * Data transformation functions for notification API responses.
 *
 * Mappers convert between:
 * - API/Supabase row format (snake_case)
 * - Domain model format (camelCase)
 *
 * This centralizes all data transformation logic for easier testing
 * and maintenance.
 */

import type { Notification, NotificationRow } from '../model';

// ============================================================================
// Response → Domain Mappers
// ============================================================================

/**
 * Map notification row to domain model.
 */
export function mapNotificationRow(row: NotificationRow): Notification {
  return {
    id: row.id,
    userId: row.user_id,
    type: row.type,
    title: row.title,
    body: row.body,
    metadata: row.metadata ?? {},
    readAt: row.read_at,
    createdAt: row.created_at,
  };
}

File: apps/backend-python/src/housing_platform/notifications/mappers.py
from housing_platform.db.models import Notifications
from housing_platform.notifications.schemas import Notification, NotificationType


def map_notification(notification: Notifications) -> Notification:
    return Notification(
        id=notification.id,
        user_id=notification.user_id,
        type=NotificationType(notification.type),
        title=notification.title,
        body=notification.body,
        metadata=notification.metadata_ or {},
        read_at=notification.read_at,
        created_at=notification.created_at,
    )
